File: ubag-python/src/ubag/_challenge.py
"""
Branch C — agent identity challenge (asymmetric).

An unknown agent is issued a one-time nonce. It proves control of its identity
key by signing the nonce with its Ed25519 PRIVATE key; the server verifies with
the agent's PUBLIC key. That is what establishes *who* the agent is — knowledge
of a shared secret never could.

The nonce carries a server-side HMAC `stamp` so the server can confirm it issued
the nonce without storing it (stateless issuance). That HMAC is the server signing
to *itself* — it is not part of the identity proof.

Replay: nonces are one-time. Provide a shared `store` (Redis/DB) in multi-process
deployments; the in-memory default only protects a single process.
"""
from __future__ import annotations

import hashlib
import heapq
import hmac
import secrets
import threading
import time
from collections import OrderedDict
from typing import Callable, Optional, Protocol

from ubag._keys import agent_id, agent_verify


class UBAGReplayStore(Protocol):
    def consume(self, identifier: str, expires_at: int) -> bool: ...


class ReplayStoreCapacityError(RuntimeError):
    """Replay protection cannot safely retain another live identifier."""

    def __init__(
        self,
        message: str,
        *,
        tier: str = "global",
        namespace: Optional[str] = None,
    ) -> None:
        super().__init__(message)
        self.tier = tier
        self.namespace = namespace


class MemoryReplayStore:
    """Bounded, thread-safe TTL replay store for one process."""

    def __init__(
        self,
        max_entries: int = 10_000,
        *,
        max_entries_per_namespace: Optional[int] = None,
        namespace_extractor: Optional[Callable[[str], str]] = None,
        max_entries_per_group: Optional[int] = None,
        group_extractor: Optional[Callable[[str], str]] = None,
    ) -> None:
        if max_entries < 1:
            raise ValueError("max_entries must be positive")
        self.max_entries = max_entries
        if max_entries_per_namespace is not None and max_entries_per_namespace < 1:
            raise ValueError("max_entries_per_namespace must be positive")
        self.max_entries_per_namespace = max_entries_per_namespace
        self._namespace_extractor = namespace_extractor
        if max_entries_per_group is not None and max_entries_per_group < 1:
            raise ValueError("max_entries_per_group must be positive")
        self.max_entries_per_group = max_entries_per_group
        self._group_extractor = group_extractor
        self._entries: OrderedDict[str, int] = OrderedDict()
        self._namespace_counts: dict[str, int] = {}
        self._group_counts: dict[str, int] = {}
        self._expiry_heap: list[tuple[int, str]] = []
        self._lock = threading.Lock()
        self.capacity_exhaustions = 0

    def consume(self, identifier: str, expires_at: int) -> bool:
        now = int(time.time())
        with self._lock:
            while self._expiry_heap and self._expiry_heap[0][0] <= now:
                expiry, key = heapq.heappop(self._expiry_heap)
                if self._entries.get(key) == expiry:
                    self._entries.pop(key, None)
                    namespace = self._namespace(key)
                    if namespace is not None:
                        remaining = max(
                            self._namespace_counts.get(namespace, 0) - 1, 0
                        )
                        if remaining:
                            self._namespace_counts[namespace] = remaining
                        else:
                            self._namespace_counts.pop(namespace, None)
                    group = self._group(key)
                    if group is not None:
                        remaining = max(self._group_counts.get(group, 0) - 1, 0)
                        if remaining:
                            self._group_counts[group] = remaining
                        else:
                            self._group_counts.pop(group, None)
            if identifier in self._entries:
                return False
            namespace = self._namespace(identifier)
            if (
                namespace is not None
                and self.max_entries_per_namespace is not None
                and self._namespace_counts.get(namespace, 0)
                >= self.max_entries_per_namespace
            ):
                self.capacity_exhaustions += 1
                raise ReplayStoreCapacityError(
                    "replay namespace is full of live entries; refusing to fail open",
                    tier="namespace",
                    namespace=namespace,
                )
            group = self._group(identifier)
            if (
                group is not None
                and self.max_entries_per_group is not None
                and self._group_counts.get(group, 0) >= self.max_entries_per_group
            ):
                self.capacity_exhaustions += 1
                raise ReplayStoreCapacityError(
                    "replay group is full of live entries; refusing to fail open",
                    tier="group",
                )
            if len(self._entries) >= self.max_entries:
                self.capacity_exhaustions += 1
                raise ReplayStoreCapacityError(
                    "replay store is full of live entries; refusing to fail open",
                    tier="global",
                )
            self._entries[identifier] = expires_at
            if namespace is not None:
                self._namespace_counts[namespace] = (
                    self._namespace_counts.get(namespace, 0) + 1
                )
            if group is not None:
                self._group_counts[group] = self._group_counts.get(group, 0) + 1
            self._entries.move_to_end(identifier)
            heapq.heappush(self._expiry_heap, (expires_at, identifier))
            # Bounded entries plus a heap gives O(log n) hot-path expiry.
            if len(self._expiry_heap) > (2 * len(self._entries) + 64):
                self._expiry_heap = [
                    (expiry, key) for key, expiry in self._entries.items()
                ]
                heapq.heapify(self._expiry_heap)
            return True

    def _namespace(self, identifier: str) -> Optional[str]:
        if self._namespace_extractor is None:
            return None
        return self._namespace_extractor(identifier)

    def _group(self, identifier: str) -> Optional[str]:
        if self._group_extractor is None:
            return None
        return self._group_extractor(identifier)


class _MemoryNonceStore(MemoryReplayStore):
    """Backward-compatible internal name."""


_default_nonce_store = MemoryReplayStore()
_default_pop_store = MemoryReplayStore()


def _stamp(server_secret: str, nonce: str, ts: int) -> str:
    return hmac.new(
        server_secret.encode(), f"{nonce}:{ts}".encode(), hashlib.sha256
    ).hexdigest()


def generate_challenge(server_secret: str, ttl: int = 120) -> dict:
    """Generate a nonce challenge for an unknown agent (sent in the 429 body).

    The agent must:
      1. read `nonce`
      2. sign the nonce bytes with its Ed25519 private key
      3. POST {nonce, timestamp, stamp, agent_public, signature} to /ubag/verify
    """
    nonce = secrets.token_urlsafe(32)
    ts = int(time.time())
    return {
        "nonce": nonce,
        "timestamp": ts,
        "ttl": ttl,
        "algo": "Ed25519",
        "stamp": _stamp(server_secret, nonce, ts),
        "instructions": (
            "Sign the `nonce` bytes with your Ed25519 private key and POST "
            "{nonce, timestamp, stamp, agent_public, signature} to /ubag/verify."
        ),
    }


def verify_challenge(
    server_secret: str,
    nonce: str,
    timestamp: int,
    stamp: str,
    agent_public: str,
    signature: str,
    ttl: int = 120,
    store: Optional[UBAGReplayStore] = None,
) -> tuple[bool, str, Optional[str]]:
    """Verify a challenge response. Returns (ok, reason, agent_id).

    1. Stamp    — the server issued this exact (nonce, ts), untampered
    2. Expiry   — the timestamp is current and within TTL
    3. Identity — the signature verifies under agent_public
    4. Replay   — the nonce is atomically consumed after verification
    """
    nonce_store = store or _default_nonce_store

    if not nonce or not agent_public or not signature:
        return False, "missing_fields", None

    if not hmac.compare_digest(_stamp(server_secret, nonce, timestamp), stamp):
        return False, "invalid_stamp", None

    now = int(time.time())
    age = now - timestamp
    if age > ttl:
        return False, "nonce_expired", None
    if age < -5:
        return False, "nonce_from_future", None

    # The identity proof: only the holder of the matching private key can produce this.
    if not agent_verify(agent_public, nonce.encode(), signature):
        return False, "bad_signature", None

    try:
        consumed = nonce_store.consume(nonce, timestamp + ttl)
    except ReplayStoreCapacityError:
        return False, "replay_store_exhausted", None
    if not consumed:
        return False, "nonce_already_used", None
    return True, "identity_verified", agent_id(agent_public)


def build_pop_message(
    method: str,
    host: str,
    target: str,
    token: str,
    ts: int,
    jti: str,
) -> bytes:
    """Build the v2 request-bound proof-of-possession message."""
    token_hash = hashlib.sha256(token.encode()).hexdigest()
    return (
        "UBAG-POP-V2\n"
        f"{method.upper()}\n{host.lower()}\n{target}\n{token_hash}\n{int(ts)}\n{jti}"
    ).encode()


def verify_pop(
    agent_public: str,
    method: str,
    host: str,
    target: str,
    token: str,
    ts: int,
    jti: str,
    signature: str,
    max_age: int = 60,
    store: Optional[UBAGReplayStore] = None,
) -> bool:
    """Verify a request-bound, one-time proof of possession.

    The proof covers method, host, path plus query, the credential thumbprint,
    timestamp, and a unique proof identifier. The identifier is atomically
    consumed after signature verification to reject within-window replay.
    """
    if not agent_public or not host or not target or not token or not jti or not signature:
        return False
    try:
        ts = int(ts)
    except (TypeError, ValueError):
        return False
    now = int(time.time())
    if abs(now - ts) > max_age:
        return False
    message = build_pop_message(method, host, target, token, ts, jti)
    if not agent_verify(agent_public, message, signature):
        return False
    replay_store = store or _default_pop_store
    try:
        return replay_store.consume(f"pop:{jti}", now + max_age)
    except ReplayStoreCapacityError:
        return False

File: ubag-node/src/challenge.js
'use strict';

/**
 * Branch C — agent identity challenge (asymmetric Ed25519).
 *
 * The agent proves control of its identity key by signing the nonce with its
 * PRIVATE key; the server verifies with the agent's PUBLIC key. The nonce carries
 * a server HMAC `stamp` for stateless issuance (the server signing to itself —
 * not part of the identity proof). The old timing/cadence gate is removed.
 *
 * Replay: provide a shared `nonceStore` in multi-process deployments; the
 * in-memory default only protects a single process.
 */
const crypto = require('crypto');
const { agentVerify, agentId } = require('./keys');

const DEFAULT_TTL = 120;

class ReplayStoreCapacityError extends Error {
  constructor(message = 'replay store is full of live entries; refusing to fail open',
    { tier = 'global', namespace = null } = {}) {
    super(message);
    this.name = 'ReplayStoreCapacityError';
    this.tier = tier;
    this.namespace = namespace;
  }
}

class MemoryReplayStore {
  constructor(maxEntries = 10000, {
    maxEntriesPerNamespace = null,
    namespaceExtractor = null,
    maxEntriesPerGroup = null,
    groupExtractor = null,
  } = {}) {
    if (!Number.isInteger(maxEntries) || maxEntries < 1) throw new Error('maxEntries must be positive');
    this.maxEntries = maxEntries;
    if (maxEntriesPerNamespace !== null &&
        (!Number.isInteger(maxEntriesPerNamespace) || maxEntriesPerNamespace < 1)) {
      throw new Error('maxEntriesPerNamespace must be positive');
    }
    this.maxEntriesPerNamespace = maxEntriesPerNamespace;
    this.namespaceExtractor = namespaceExtractor;
    if (maxEntriesPerGroup !== null &&
        (!Number.isInteger(maxEntriesPerGroup) || maxEntriesPerGroup < 1)) {
      throw new Error('maxEntriesPerGroup must be positive');
    }
    this.maxEntriesPerGroup = maxEntriesPerGroup;
    this.groupExtractor = groupExtractor;
    this.entries = new Map();
    this.namespaceCounts = new Map();
    this.groupCounts = new Map();
    this.expiryHeap = [];
    this.capacityExhaustions = 0;
  }

  _less(a, b) {
    return a[0] < b[0] || (a[0] === b[0] && a[1] < b[1]);
  }

  _push(node) {
    const heap = this.expiryHeap;
    heap.push(node);
    let index = heap.length - 1;
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (!this._less(heap[index], heap[parent])) break;
      [heap[index], heap[parent]] = [heap[parent], heap[index]];
      index = parent;
    }
  }

  _pop() {
    const heap = this.expiryHeap;
    const first = heap[0];
    const last = heap.pop();
    if (heap.length && last) {
      heap[0] = last;
      let index = 0;
      while (true) {
        const left = index * 2 + 1;
        const right = left + 1;
        let smallest = index;
        if (left < heap.length && this._less(heap[left], heap[smallest])) smallest = left;
        if (right < heap.length && this._less(heap[right], heap[smallest])) smallest = right;
        if (smallest === index) break;
        [heap[index], heap[smallest]] = [heap[smallest], heap[index]];
        index = smallest;
      }
    }
    return first;
  }

  consume(identifier, expiresAt) {
    const now = Math.floor(Date.now() / 1000);
    while (this.expiryHeap.length && this.expiryHeap[0][0] <= now) {
      const [expiry, key] = this._pop();
      if (this.entries.get(key) === expiry) {
        this.entries.delete(key);
        const namespace = this._namespace(key);
        if (namespace !== null) {
          const remaining = Math.max((this.namespaceCounts.get(namespace) || 0) - 1, 0);
          if (remaining) this.namespaceCounts.set(namespace, remaining);
          else this.namespaceCounts.delete(namespace);
        }
        const group = this._group(key);
        if (group !== null) {
          const remaining = Math.max((this.groupCounts.get(group) || 0) - 1, 0);
          if (remaining) this.groupCounts.set(group, remaining);
          else this.groupCounts.delete(group);
        }
      }
    }
    if (this.entries.has(identifier)) return false;
    const namespace = this._namespace(identifier);
    if (namespace !== null && this.maxEntriesPerNamespace !== null &&
        (this.namespaceCounts.get(namespace) || 0) >= this.maxEntriesPerNamespace) {
      this.capacityExhaustions += 1;
      throw new ReplayStoreCapacityError(
        'replay namespace is full of live entries; refusing to fail open',
        { tier: 'namespace', namespace }
      );
    }
    const group = this._group(identifier);
    if (group !== null && this.maxEntriesPerGroup !== null &&
        (this.groupCounts.get(group) || 0) >= this.maxEntriesPerGroup) {
      this.capacityExhaustions += 1;
      throw new ReplayStoreCapacityError(
        'replay group is full of live entries; refusing to fail open',
        { tier: 'group' }
      );
    }
    if (this.entries.size >= this.maxEntries) {
      this.capacityExhaustions += 1;
      throw new ReplayStoreCapacityError(undefined, { tier: 'global' });
    }
    this.entries.set(identifier, expiresAt);
    if (namespace !== null) {
      this.namespaceCounts.set(namespace, (this.namespaceCounts.get(namespace) || 0) + 1);
    }
    if (group !== null) {
      this.groupCounts.set(group, (this.groupCounts.get(group) || 0) + 1);
    }
    this._push([expiresAt, identifier]);
    if (this.expiryHeap.length > 2 * this.entries.size + 64) {
      this.expiryHeap = [];
      for (const [key, expiry] of this.entries) this._push([expiry, key]);
    }
    return true;
  }

  _namespace(identifier) {
    if (typeof this.namespaceExtractor !== 'function') return null;
    return this.namespaceExtractor(identifier);
  }

  _group(identifier) {
    if (typeof this.groupExtractor !== 'function') return null;
    return this.groupExtractor(identifier);
  }
}

const defaultNonceStore = new MemoryReplayStore();
const defaultPopStore = new MemoryReplayStore();

function stamp(serverSecret, nonce, ts) {
  return crypto.createHmac('sha256', serverSecret).update(`${nonce}:${ts}`).digest('hex');
}

function generateChallenge(serverSecret, ttl = DEFAULT_TTL) {
  const nonce = crypto.randomBytes(32).toString('base64url');
  const timestamp = Math.floor(Date.now() / 1000);
  return {
    nonce,
    timestamp,
    ttl,
    algo: 'Ed25519',
    stamp: stamp(serverSecret, nonce, timestamp),
    instructions:
      'Sign the `nonce` bytes with your Ed25519 private key and POST ' +
      '{nonce, timestamp, stamp, agent_public, signature} to /ubag/verify.',
  };
}

function verifyChallenge(
  serverSecret,
  { nonce, timestamp, stamp: stampIn, agent_public, signature },
  { ttl = DEFAULT_TTL, nonceStore = null } = {}
) {
  const store = nonceStore || defaultNonceStore;

  if (!nonce || !agent_public || !signature) return [false, 'missing_fields', null];

  const expected = stamp(serverSecret, nonce, timestamp);
  const a = Buffer.from(expected);
  const b = Buffer.from(String(stampIn));
  if (a.length !== b.length || !crypto.timingSafeEqual(a, b)) return [false, 'invalid_stamp', null];

  const now = Math.floor(Date.now() / 1000);
  const age = now - timestamp;
  if (age > ttl) return [false, 'nonce_expired', null];
  if (age < -5) return [false, 'nonce_from_future', null];

  // The identity proof: only the holder of the matching private key can produce this.
  if (!agentVerify(agent_public, nonce, signature)) return [false, 'bad_signature', null];

  try {
    if (!store.consume(nonce, timestamp + ttl)) return [false, 'nonce_already_used', null];
  } catch (error) {
    if (error instanceof ReplayStoreCapacityError) {
      return [false, 'replay_store_exhausted', null];
    }
    throw error;
  }
  return [true, 'identity_verified', agentId(agent_public)];
}

function buildPopMessage(method, host, target, token, ts, jti) {
  const tokenHash = crypto.createHash('sha256').update(String(token)).digest('hex');
  return [
    'UBAG-POP-V2',
    String(method).toUpperCase(),
    String(host).toLowerCase(),
    String(target),
    tokenHash,
    String(ts),
    String(jti),
  ].join('\n');
}

/**
 * V2 proof-of-possession for a credentialed request. The proof binds method,
 * host, path+query, credential thumbprint, timestamp, and a one-time identifier.
 */
function verifyPop(
  agentPublic,
  method,
  host,
  target,
  token,
  ts,
  jti,
  signature,
  { maxAge = 60, replayStore = defaultPopStore } = {}
) {
  if (!agentPublic || !host || !target || !token || !jti || !signature) return false;
  const t = parseInt(ts, 10);
  if (!Number.isFinite(t)) return false;
  const now = Math.floor(Date.now() / 1000);
  if (Math.abs(now - t) > maxAge) return false;
  const message = buildPopMessage(method, host, target, token, t, jti);
  if (!agentVerify(agentPublic, message, signature)) return false;
  try {
    return replayStore.consume(`pop:${jti}`, now + maxAge);
  } catch (error) {
    if (error instanceof ReplayStoreCapacityError) return false;
    throw error;
  }
}

module.exports = {
  ReplayStoreCapacityError,
  MemoryReplayStore,
  generateChallenge,
  verifyChallenge,
  buildPopMessage,
  verifyPop,
  stamp,
};
